cpp/src/arrow/io/file.h

```diff
@@ -80,6 +80,56 @@ class ARROW_EXPORT FileOutputStream : public OutputStream {
   std::unique_ptr<FileOutputStreamImpl> impl_;
 };
 
+#if defined(__linux__)
+/// \brief An operating system file open in write-only and O_DIRECT mode, only works on
+/// Linux.
+class ARROW_EXPORT DirectFileOutputStream : public OutputStream {
+ public:
+  ~DirectFileOutputStream() override;
+
+  /// \brief Open a local file for writing, truncating any existing file
+  /// \param[in] path with UTF8 encoding
+  ///
+  /// When opening a new file, any existing file with the indicated path is
+  /// truncated to 0 bytes, deleting any existing data
+  static Result<std::shared_ptr<DirectFileOutputStream>> Open(const std::string& path);
+
+  /// \brief Open a file descriptor for writing.  The underlying file isn't
+  /// truncated.
+  /// \param[in] fd file descriptor
+  /// \param[in] sector_size the physical disk sector size hosting this fd
+  /// \return an open FileOutputStream
+  ///
+  /// The file descriptor becomes owned by the OutputStream, and will be closed
+  /// on Close() or destruction.
+  static Result<std::shared_ptr<DirectFileOutputStream>> Open(int fd,
+                                                              int64_t sector_size);
+
+  // OutputStream interface
+  Status Close() override;
+  bool closed() const override;
+  Result<int64_t> Tell() const override;
+
+  // Write bytes to the stream. Thread-safe
+  Status Write(const void* data, int64_t nbytes) override;
+  /// \cond FALSE
+  using Writable::Write;
+  /// \endcond
+
+  int file_descriptor() const;
+
+ private:
+  explicit DirectFileOutputStream(int64_t sector_size);
+
+  class ARROW_NO_EXPORT DirectFileOutputStreamImpl;
+  std::unique_ptr<DirectFileOutputStreamImpl> impl_;
+  int64_t sector_size_ = 0;
+  std::vector<uint8_t> cached_data_;
+  uint8_t* aligned_cached_data_;
+  int64_t cached_length_ = 0;
```

Comment:
Let's put this inside the private impl class instead of here?

Reply:
What's the problem with having these attributes here instead of inside the private impl class?

I don't think it makes sense to have them inside the private impl class, because they are used by the DirectFileOutputStream class methods not the impl_ class methods. The impl_ class methods just override basic OSFile functionality and should not have to know about cached_data etc.